Update wording, links in key projects page

brainwane opened the pull request:
A few link fixes, clearer explanations regarding wheel, packaging, pip, and distlib, and a few wording/capitalization fixes.

webknjaz commented on source/key_projects.rst:
```diff
@@ -435,7 +454,7 @@ distutils
 User irc:#pypa  |
 Dev irc:#pypa-dev
 
-A package in the Python Standard Library that has support for creating and
```
Technically it's a package https://github.com/python/cpython/tree/master/Lib/distutils
(module == one py file, package == dir)

Also, this change is inconsistent with the paragraph about ensurepip above.

brainwane replied:
Thanks. Resolved.

ncoghlan commented on source/key_projects.rst:
```diff
@@ -54,13 +59,20 @@ packaging
+without needing to explicitly call it; developers of the other Python
+packaging, distribution, and installation tools listed here often use
+its functionality to parse, discover, and otherwise handle dependency
+attributes.
```
```suggestion
attributes.

This project specifically focuses on implementing the modern Python packaging
interoperability standards defined at :ref:`packaging-specifications`, and will report
errors for sufficiently old legacy packages that are incompatible with those
standards. The :ref:`distlib` project is a more permissive library that attempts to
provide a plausible reading of ambiguous metadata in cases where ``packaging``
will instead report on error.
```

It seems worth highlighting that the essential difference between "packaging" and "distlib" is that the latter includes extra behaviour that isn't in the approved PEPs, and one of the key aspects of that is making more guesses when confronted with ambiguous legacy metadata.

brainwane replied:
Thanks so much! Added.

ncoghlan commented on source/key_projects.rst:
```diff
@@ -71,12 +83,13 @@ pip
 
-A tool for installing Python packages.
+The most popular tool for installing Python packages, and the one
+included with modern versions of Python.
```
```suggestion
included with modern versions of Python.

It provides the essential core features for finding, downloading, and installing packages
from pypi.org and other index servers, and can be incorporated into a wide range of
development workflows via its CLI interface.
```

brainwane replied:
Thank you! Added.

ncoghlan commented on source/key_projects.rst:
```diff
@@ -150,7 +163,7 @@ twine
+`GitHub <https://github.com/pypa/twine>`__ |
 `PyPI <https://pypi.org/project/twine>`__
 
 Twine is a utility for interacting with PyPI, that offers a secure replacement for
```
We should update this description, as ``setup.py upload`` is secure in modern versions of Python (since 3.5.3 or so).

``twine`` is still better - just for reasons other than this one.

brainwane replied:
I updated it more thoroughly. Thanks.

ncoghlan commented on source/key_projects.rst:
```diff
@@ -435,7 +454,7 @@ distutils
 Dev irc:#pypa-dev
 
-A package in the Python Standard Library that has support for creating and
+A module in the Python Standard Library that has support for creating and
```
```suggestion
The original Python packaging system, added to the standard library in Python 2.0.
Due to the challenges of maintaining a packaging system where feature updates are tightly coupled to language runtime updates, direct usage of ``distutils`` is now actively discouraged, with :ref:`Setuptools` being the preferred replacement. :ref:`Setuptools` not only provides features that plain ``distutils`` doesn't offer (such as dependency declarations and entry point declarations), it also provides a consistent build interface and feature set across all supported Python versions.
```

brainwane replied:
Thanks so much! Incorporated your addition.

brainwane commented on source/key_projects.rst:
```diff
@@ -39,12 +39,17 @@ distlib
+to packaging and distribution of Python software.  It consists in part
+of the functions from the `distutils2
+<https://pypi.org/project/Distutils2>`_ project, which was intended to
+be released as ``packaging`` in the Python 3.3 stdlib, but was removed
```
Thanks! Fixed, I believe..

brainwane commented on source/key_projects.rst:
```diff
@@ -39,12 +39,17 @@ distlib
+useful for developers of third-party packaging tools to make and
+upload binary and source distributions, achieve interoperability,
+resolve dependencies, manage package resources, and do other similar
+functions.
```
Thanks - added.

brainwane commented:
> Several suggestions inline. Perhaps it would make sense to split the purely mechanical GitHub word mark fix out from the more subjective updates to the project descriptions?

I appreciate the note; if I'd been more disciplined I would have done that! If this PR drags on for several days then I'll probably do that. :-)

bhrutledge commented on source/key_projects.rst:
```diff
@@ -150,12 +180,14 @@ twine
-``setup.py upload``.
-
+Twine is the primary tool developers use to upload packages to the
+Python Package Index or other Python package indexes. It is a
```
```suggestion
PyPI or other Python package indexes. It is a
```

bhrutledge commented on source/key_projects.rst:
```diff
@@ -71,13 +95,18 @@ pip
+included with modern versions of Python.
 
+It provides the essential core features for finding, downloading, and
+installing packages from pypi.org and other index servers, and can be
```
This feels more in line with other references to these concepts:

```suggestion
installing packages from PyPI and other Python package indexes, and can be
```

bhrutledge commented on source/key_projects.rst:
```diff
@@ -150,12 +180,14 @@ twine
-``setup.py upload``.
-
+Twine is the primary tool developers use to upload packages to the
+Python Package Index or other Python package indexes. It is a
```
How about ths, to avoid the repetition of "package index"?

```
Twine is the primary tool developers use to upload packages to
PyPI or other Python package indexes. It is a
```

It looks like the acronym is more common in this document than the words.

brainwane replied:
It is. I'm guessing a lot of people come to section anchor tags in this document and skim around, and I think it's okay to say the full name once in a while. I'm not going to figure out a consistent style guide and apply it in this particular PR, but if someone else does so in a later change and it changes my choices in this regard, then I will be fine with that; I will not cry out to the heavens and rend my clothes or anything,

bhrutledge commented on source/key_projects.rst:
```diff
@@ -348,19 +386,59 @@ pipx
+`GitHub <https://github.com/piwheels/piwheels/>`__
+
+Piwheels is a website, and software underpinning it, that fetches
+source code distribution packages from the Python Package Index and
```
```suggestion
source code distribution packages from PyPI and
```

Similar to my comment re: Twine

bhrutledge commented on source/key_projects.rst:
```diff
@@ -348,19 +386,59 @@ pipx
 ====
 
 `Docs <https://github.com/pipxproject/pipx>`__ |
```
```suggestion
`Docs <https://pipxproject.github.io/pipx/>`__ |
```

webknjaz replied:
This suggested change breaks indentation :point_up: (which is a known bug @ GH)
Here's a fixed suggestion:
```suggestion
 `Docs <https://pipxproject.github.io/pipx/>`__ |
```

bhrutledge commented on source/key_projects.rst:
```diff
@@ -54,13 +62,29 @@ packaging
 
 Core utilities for Python packaging used by :ref:`pip` and :ref:`setuptools`.
 
+The core utilities in the packaging bundle handle version handling,
```
The word "bundle" surprised me; I hadn't seen it used in the context of Python packaging. Wondering if "library" is more accurate?

```suggestion
The core utilities in the packaging library handle version handling,
```

bhrutledge commented on source/key_projects.rst:
```diff
@@ -54,13 +62,29 @@ packaging
 
+The core utilities in the packaging bundle handle version handling,
+specifiers, markers, requirements, tags, and similar attributes and
+tasks for Python packages. Most Python users rely on this bundle
```
```suggestion
tasks for Python packages. Most Python users rely on this library
```

bhrutledge commented on source/key_projects.rst:
```diff
@@ -54,13 +62,29 @@ packaging
+sufficiently old legacy packages that are incompatible with those
+standards. In contrast, the :ref:`distlib` project is a more
+permissive library that attempts to provide a plausible reading of
+ambiguous metadata in cases where ``packaging`` will instead report on
```
Not sure what the convention is (or should be) around monospacing package names. It seems like most package names are *not* monospaced, except here, once in "Distlib", and in "distutils".

brainwane replied:
Yeah, I think there is no consistent convention. Future person who decides on a style guide, consider this question!

bhrutledge commented:
Also, thanks for putting the time into this, and the attention to detail. I think it’s a nice improvement.

webknjaz commented on source/key_projects.rst:
```diff
@@ -407,6 +485,19 @@ Spack is not in PyPI (yet), but it requires no installation and can be
+`GitHub <https://github.com/zestsoftware/zest.releaser/>`__ |
+`PyPI <https://pypi.org/project/zest.releaser/>`__
+
+zest.releaser is a Python package release tool providing an
```
How about this?
```suggestion
``zest.releaser`` is a Python package release tool providing an
```
It seems like there's no consistency for marking dist names with `<code>` on this page. You added it for distutils so maybe others deserve this too.

webknjaz commented on source/key_projects.rst:
```diff
@@ -348,19 +386,59 @@ pipx
+compiles them into binary wheels that are optimized for installation
+onto Raspberry Pi computers. Many Raspberry Pi users who use or
+develop Python tools to run on Raspbian use piwheels as their primary
+Python package index.
```
Recently, I've discovered that this index doesn't support PEP 503 which makes Pip explode for wheels that specify `Requires-Python` under interpreter versions that don't match the condition instead of selecting wheels properly on the download stage. (https://github.com/cherrypy/cherrypy/issues/1837)

I think this may be a good place to add a known issue warning about this.

brainwane replied:
I will do so; I'll link to https://github.com/piwheels/piwheels/issues/208 which is where they are fixing it. Thanks for the heads-up.

brainwane commented:
Added several projects and improved wording and detail for several existing projects.

webknjaz commented on source/key_projects.rst:
```diff
@@ -347,20 +443,82 @@ make deployment of Python applications as simple as ``cp``.
+computers. Many Raspberry Pi users who use or develop Python tools to
+run on Raspbian use Piwheels as their primary Python package index.
+
+(Note that Piwheels `does not yet fully support
```
This should probably be a warning box marked in RST...

webknjaz replied:
```suggestion
  .. warning::

    Note that Piwheels `does not yet fully support
```

webknjaz commented on source/key_projects.rst:
```diff
@@ -347,20 +443,82 @@ make deployment of Python applications as simple as ``cp``.
+
+(Note that Piwheels `does not yet fully support
+<https://github.com/piwheels/piwheels/issues/208>`__ :pep:`503` and
+thus some wheels have trouble with inconsistent ``Requires-Python``
```
This sounds wrong. Wheels don't have troubles, they are correct, it's Pip that has troubles because it does not get enough info from the index and cannot properly identify what to start downloading.

webknjaz replied:
I'd maybe go as far as pointing to pip-tools as a workaround mentioning that this will help end-users pin versions of transitive deps.

brainwane replied:
At this point I'm going to write a very simplified line of explanation here and I suggest you make a separate PR or commit with the language that you prefer. :-)

pradyunsg commented on source/key_projects.rst:
```diff
@@ -167,12 +221,18 @@ virtualenv
-A tool for creating isolated Python environments.
+virtualenv is a tool which uses the command-line path environment
+variable to (like venv) create isolated Python :term:`Virtual
+Environments <Virtual Environment>`. virtualenv goes beyond venv by
```
```
virtualenv is a tool to create isolated Python :term:`Virtual
Environments <Virtual Environment>` (like venv). virtualenv provides
additional functionality compared to venv by...
```

brainwane replied:
Thanks. Fixed.

pradyunsg commented on source/key_projects.rst:
```diff
@@ -199,16 +259,22 @@ wheel
 creating :term:`wheel distributions <Wheel>`.  Additionally, it offers its own
 command line utility for creating and installing wheels.
 
+See also `auditwheel <https://github.com/pypa/auditwheel>`__, a tool
```
Any specific reason to not make auditwheel a dedicated heading?

brainwane replied:
Two reasons.

1. `auditwheel` is, as I understand it, only useful in the context of `wheel`.
2. Alphabetically, they're super far from each other.

But I don't particularly care. If you would like to block this PR on that, please say so; otherwise, I figure if someone would like to do that, they can do so and I will not object. :-)

webknjaz replied:
This is questionable: technically, it's useful for making manylinux wheels but `wheel` dist isn't used directly. You could create a wheel using any tool you like and then use `auditwheel` to inspect it and/or make it manylinux compliant. Most of the guides just use `pip wheel` command to make that first step. So whether it's only useful in the context of `wheel` probably depends on the PoV.

ncoghlan replied:
I was going to comment on this in my initial review, but came to the same conclusion as @brainwane: with the current layout of this page, giving auditwheel its own section would look quite odd, whereas it fits better here.

If we were to give it a top level section, I'd suggest doing it in a separate PR where we categorise this list by project type (e.g. installation, build & publishing, package hosting, utility library) rather than splitting it out here.

pradyunsg replied:
> If you would like to block this PR on that, please say so

Nope. :)

uranusjr commented:
I’m on my phone right now; there are several casing and format inconsistencies that could use some fixing. I see at least three different rendering in distlib alone: Distlib, distlib, and `distlib`.

brainwane commented:
I've fixed up a few formatting inconsistencies (addressing @uranusjr's comments) as best I could given the inconsistency of the formatting beforehand. Also I've now added web links to the IRC channels.

I believe this is ready to merge and would substantially improve the Key Projects page when merged, with better basic overviews of many important packaging projects. I respect that this PR is a chance to fix a bunch of formatting and wording stuff in a piece of documentation that hasn't gotten a ton of love lately. May I ask that we try to merge this is, and make follow-up PRs or issues about further improving casing/formatting consistency, advising users in more detail about gotchas, and so on?
